# epithelial_cell_diversity

The following code was used to analyse different types of omics-data.

The analysis is separated into different parts:

* Alignment using STAR
* Genome Mapping of bulk data using featurecounts
* Bulk-RNA-Seq analysis using DESeq2
	* Bulk analysis full small intestine only uninfected
	* Bulk analysis full small intestine including infected
	* Bulk analysis crypts vs villi
 	* Bulk analysis proximal medial distal
* Proteomics analysis using limma
* Single cell analysis using Seurat
	* Costalab single cell pipeline
 	* Single cell sub analyses
  * Transcription factor analysis using Dorothea
