---
title: "DE_GO_RNA_seq"
author: "Johannes Schöneich"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(EnhancedVolcano)
library(tidyverse)
library(DESeq2)
library(limma)
library(glue)
library(clusterProfiler)
library(DOSE)
library(enrichplot)

harddrive <- "H:"

source(file.path(harddrive,"helper_functions.R"))
source(file.path(harddrive,"DE_analysis/DE_helperfunctions.R"))
```

```{r}
dds_fullsample <- readRDS(file.path(harddrive,
          "RNA_seq/RNAseq_uninfected/dds_LRT_RNAseq_uninfected.Rds"))

#condition to analyse: e.g. timepoint, condition
condition <- "timepoint"

analysis_folder <- "DE_analysis/DE_LRT_uninfected"

plots_folder <- file.path(harddrive,analysis_folder,"DE/plots")
charts_folder <- file.path(harddrive,analysis_folder,"DE/charts")

#if (!dir.exists(analysis_folder)){
  dir.create(analysis_folder,recursive = TRUE)
  dir.create(plots_folder,recursive = TRUE)
  dir.create(charts_folder,recursive = TRUE)
#}

# optional: read in the results.list if this analysis was already run 
# and you want to change plots
results.list <- readRDS(file.path(harddrive,analysis_folder,
                                  glue("DE_results.list_{condition}.Rds")))

contribution.df <- read.xlsx(file.path(harddrive,"RNA_seq/RNAseq_uninfected/",
                                       "pca_driver_genes_RNAseq_uninfected.xlsx"),
                             rowNames = TRUE)

#we only need the first 100, they drive the biplot
contribution.df <- contribution.df[1:100,]
```

## Volcano Plots DE genes

The contrast order is always +logFC then -logFC 
(first numerator (Zähler) then Denominator (Nenner))
condition treated vs untreated, tells you that the estimates are of the 
logarithmic fold change log2(treated/untreated).

The level given last is the base level for the comparison. The syntax is:
contrast <- c("condition", "level_to_compare", "base_level")
for example, if we observe a log2 fold change of -2 this would mean 
the gene expression is lower in the level of interest relative to the base level

```{r, warning=FALSE}
#every pairwise combination of timepoints
combination <- combn(levels(colData(dds_fullsample)[,condition]), 2)
list.names <- capture.output(apply(combination,2,function(x){cat(x,sep = "_",fill = TRUE)}))
list.names <- list.names[which(list.names!="NULL")]

if(!exists("results.list")){
#make an empty list with all combinations
results.list <- sapply(list.names,function(x) NULL)

for(i in 1:length(results.list)){
  print(i)
  #paste("working on", list.names[i])                                     
  results.list[[i]] <- results(dds_fullsample, 
                               contrast = c(condition, combination[1,i], #d1
                                            combination[2,i]), #d5
                               cooksCutoff = TRUE, 
                               independentFiltering = TRUE,
                               alpha = 0.05,
                               pAdjustMethod = "BH",
                               test = "LRT")

  results.list[[i]] <- lfcShrink(dds_fullsample, 
                                 contrast = c(condition, combination[1,i], 
                                              combination[2,i]),
                       res = results.list[[i]],
                       type = "ashr")
}

results.list <- lapply(results.list,function(x){x$symbol <- rownames(x);return(x)})
sapply(results.list,function(x){sum(is.na(x))}) #are there any NAs left?
results.list <- lapply(results.list,function(x){na.omit(x)})

#change logfc so that early is to the left.
results.list <- lapply(results.list,
                       function(res){
                         res$log2FoldChange <- -res$log2FoldChange
                         return(res)
                         }
                       )

saveRDS(results.list,file =
          file.path(analysis_folder,glue("DE_results.list_{condition}.Rds")
                    )
        )
}
```

```{r}
plot.list <- generate_volcano_plots(results_list = results.list[c(1,4,6)],
                       pos_color = "#E41A1C",
                       neg_color = "#377EB8",
                       plots_folder = plots_folder,
                       charts_folder = charts_folder,
                       filename_prefix = "bulk_top_bot20_no_connect",
                       contribution.df
                       )

saveRDS(plot.list,file.path(harddrive,"figures/figure_1/saves/volcanos_fig1.Rds"))
```

```{r fig.height = 10, fig.width = 10}

pos_color <- "#E41A1C"
neg_color <- "#377EB8"
quantile_values <- sapply(results.list,function(x){
  quantile(x$log2FoldChange,c(0.001,0.999))
  }
)

xmin <- min(quantile_values) + 1
xmax <- max(quantile_values) - 1

p.cutoff = min(sapply(results.list,function(x){
  quantile(x[x$padj >0,]$padj,0.001)}))

for(i in 1:length(results.list)){
  res <- results.list[[i]]
  res.new <- as.data.frame(res)
  
  # extract the names from the results.list
  dayneg <- strsplit2(names(results.list[i]),split = "_")[1]
  daypos <- strsplit2(names(results.list[i]),split = "_")[2]
  
  # identify genes that pass the threshold
  up <- rownames(subset(res, log2FoldChange >= xmax))
  down <- rownames(subset(res, log2FoldChange <= xmin))
  p.up <- rownames(subset(res, padj <= p.cutoff))
  
  # change log2FCs in the object to be max or min cutoff 
  res.new$log2FoldChange <- ifelse(res.new$log2FoldChange >= xmax, xmax,
                                   ifelse(res.new$log2FoldChange <= xmin, xmin,
                                          res.new$log2FoldChange))
  # also change p.adj accordingly
  res.new$padj <- ifelse(res.new$padj <= p.cutoff, p.cutoff, res.new$padj)

  #top 20
  top20 <- res.new %>% top_n(20,log2FoldChange) %>% arrange(-log2FoldChange)
  
  #bot 20
  bot20 <- res.new %>% slice_min(log2FoldChange,n = 20) %>% arrange(-log2FoldChange)
  
  #Only those which are significant
  top20 <- subset(top20, subset = padj < 0.05 & log2FoldChange > 1) 
  bot20 <- subset(bot20, subset = padj < 0.05 & log2FoldChange < - 1)
    
  # custom shapes for the points
  customshape <- rep(19, nrow(res.new))
  names(customshape) <- rep('normal', nrow(res.new))
  customshape[which(rownames(res.new) %in% up)] <- -9658 #shape of the arrow facing right
  names(customshape)[which(rownames(res.new) %in% up)] <- '> LFC'
  customshape[which(rownames(res.new) %in% p.up)] <- 17 #shape of the arrow facing up
  names(customshape)[which(rownames(res.new) %in% p.up)] <- 'padj'
  customshape[which(rownames(res.new) %in% down)] <- -9668 #shape of the arrow facing left
  names(customshape)[which(rownames(res.new) %in% down)] <- '< LFC'
  
  # custom sizes for the points
  customsize <- rep(2.0, nrow(res.new))
  customsize [which(rownames(res.new) %in% up)] <- 4
  customsize [which(rownames(res.new) %in% p.up)] <- 4
  customsize [which(rownames(res.new) %in% down)] <- 4
  
  keyvals <- ifelse(
    res.new$log2FoldChange <= -1 & res.new$padj <= 0.05, neg_color,
    ifelse(res.new$log2FoldChange  >= 1 & res.new$padj <= 0.05, pos_color,
           'grey'))
  
  keyvals[is.na(keyvals)] <- 'grey'
  names(keyvals)[keyvals == neg_color] <- glue('up in {dayneg}')
  names(keyvals)[keyvals == 'grey'] <- 'ns'
  names(keyvals)[keyvals == pos_color] <- glue("up in {daypos}")

  # data.frame is needed for dplyr filter to work
  df <- as.data.frame(res[c("log2FoldChange","symbol","padj")]) %>% arrange(-log2FoldChange)

  up_genes <- nrow(df %>% filter(log2FoldChange >= 1 & padj <= 0.05) )
  down_genes <- nrow(df %>% filter(log2FoldChange <= -1 & padj <= 0.05))
  
  contrib_genes <- as.data.frame(df) %>% 
    filter(symbol %in% rownames(contribution.df)) %>% filter(
      abs(log2FoldChange) >= 1 & padj <= 0.05) %>% pull(symbol)
  
  plt <- EnhancedVolcano(res.new,
                         lab = res.new$symbol,
                         x = 'log2FoldChange',
                         y = 'padj',
                         xlab = bquote(~Log[2]~'fold change'),
                         ylab = bquote(~-Log[10]~'adjusted p-value'),
                         #selectLab = c(bot20$symbol,top20$symbol),
                         selectLab = contrib_genes,
                         drawConnectors = TRUE,
                         widthConnectors = 1,
                         arrowheads = FALSE,
                         max.overlaps = 5,
                         colCustom = keyvals,
                         pCutoff = 0.05,
                         FCcutoff = 1.0,
                         title = glue("DE genes {daypos} vs. {daypos}"),
                         boxedLabels = FALSE,
                         colAlpha = 0.5,
                         labSize = 6,
                         axisLabSize = 14,
                         titleLabSize = 14,
                         subtitleLabSize = 12,
                         captionLabSize = 12,
                         legendLabSize = 12,
                         #legendPosition = "right",
                         shapeCustom = customshape,
                         pointSize = customsize,
                         subtitle = glue("{dayneg}:{down_genes} {daypos}:{up_genes}"),
                         #caption = paste0("total = ", nrow(res), " genes"),
                         caption = NULL,
                         xlim = c(xmin,xmax)
  )

  save_ggplot_formats(
    plt = plt,
    base_plot_dir = plots_folder,
    plt_name = glue("Volcano_{dayneg}_vs_{daypos}_quantile_cutoff",
                    "_{condition}_5_overlap2"),
    width = 9, height = 8
  )
  
 df2 <- df %>% filter(log2FoldChange >= 1 & padj <= 0.05 |
                        log2FoldChange <= -1 & padj <= 0.05)
 write.csv(df2,file = file.path(charts_folder,
                    glue("DE_fullsamples_{dayneg}_vs_{daypos}_{condition}.csv")),
           row.names = FALSE)
}
```
