---
title: "DE_GO_RNA_seq"
author: "Johannes Schöneich"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(EnhancedVolcano)
library(tidyverse)
library(DESeq2)
library(limma)
library(glue)
library(clusterProfiler)
library(DOSE)
library(enrichplot)
library(openxlsx)
library(patchwork)

harddrive <- "D:/231248/NTFS_3.64TB"

source(file.path(harddrive,"save_load_helper.R"))
source(file.path(harddrive,"helper_functions.R"))
source(file.path(harddrive,"DE_analysis/DE_helperfunctions.R"))
```

```{r}
dds_fullsample <- readRDS(file.path(harddrive,
          "RNA_seq/RNAseq_GF/dds_LRT_RNAseq_GF.Rds"))

#condition to analyse: e.g. timepoint, condition
condition <- "timepoint"

analysis_folder <- "DE_analysis/DE_LRT_GF"

plots_folder <- file.path(harddrive,analysis_folder,"DE/plots")
charts_folder <- file.path(harddrive,analysis_folder,"DE/charts")

if (!dir.exists(analysis_folder)){
dir.create(analysis_folder,recursive = TRUE)
dir.create(plots_folder,recursive = TRUE)
dir.create(charts_folder,recursive = TRUE)
}

#optional: read in the results.list if this analysis was already run and you want to change plots
results.list <- readRDS(file.path(harddrive,glue("DE_analysis/DE_LRT_uninfected/",
                                                 "DE_results.list_{condition}.Rds")))

results.list.GF <- readRDS(file.path(harddrive,analysis_folder,
                                  glue("DE_results.list_{condition}.Rds")))

names(results.list.GF) <- c("d01GF_d05GF","d01GF_d10GF","d05GF_d10GF")

results.list.comb <- c(results.list[c(1,4,6)],results.list.GF[c(1,3)])

```

## Volcano Plots DE genes

The contrast order is always +logFC then -logFC 
(first numerator (Zähler) then Denominator (Nenner))
condition treated vs untreated, tells you that the estimates are of the 
logarithmic fold change log2(treated/untreated).

The level given last is the base level for the comparison. The syntax is:
contrast <- c("condition", "level_to_compare", "base_level")
for example, if we observe a log2 fold change of -2 this would mean the gene expression is lower 
in the level of interest relative to the base level

```{r, warning=FALSE}
# every pairwise combination of timepoints
combination <- combn(levels(colData(dds_fullsample)[,condition]), 2)
list.names <- capture.output(apply(combination,2,function(x){cat(x,sep = "_",fill = T)}))
list.names <- list.names[which(list.names!="NULL")]

if(!exists("results.list")){
  #make an empty list with all combinations
  results.list <- sapply(list.names,function(x) NULL)
  
  for(i in 1:length(results.list)){
    print(i)
    #paste("working on", list.names[i])
    results.list[[i]] <- results(dds_fullsample, contrast = c(condition, combination[1,i], 
                                                            combination[2,i]),
    cooksCutoff = TRUE, independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = "BH")
  
    results.list[[i]] <- lfcShrink(dds_fullsample, 
                                   contrast = c(condition, combination[1,i], 
                                                              combination[2,i]), 
                         res = results.list[[i]],type = "ashr")
  }
  
  results.list <- lapply(results.list,function(x){x$symbol <- rownames(x);return(x)})
  sapply(results.list,function(x){sum(is.na(x))}) #are there any NAs left?
  results.list <- lapply(results.list,function(x){na.omit(x)})
  
  #change logfc so that early is to the left.
  results.list <- lapply(results.list,
                         function(res){
                           res$log2FoldChange <- -res$log2FoldChange
                           return(res)
                           }
                         )

  saveRDS(results.list,
          file = file.path(analysis_folder,glue("DE_results.list_{condition}.Rds")))
}
```

```{r fig.height = 10, fig.width = 10}
generate_volcano_plots(results_list = results.list.comb,
                       pos_color = "#E41A1C",
                       neg_color = "#377EB8",
                       plots_folder = plots_folder,
                       charts_folder = charts_folder,
                       filename_prefix = "SPF_GF",
                       do_save = TRUE
                       )
```

```{r}
results.list.comb.filter <- lapply(results.list.comb, function(x){x %>%
    as.data.frame %>% filter(abs(log2FoldChange) >= 1 & padj<= 0.05)
  }
  )

d01_gf <- results.list.comb.filter$d01GF_d05GF %>% filter(log2FoldChange <= - 1) %>% pull(symbol) 

d01_spf <- results.list.comb.filter$d01_d05 %>% filter(log2FoldChange <= - 1) %>% pull(symbol) 

d05_gf <- results.list.comb.filter$d01GF_d05GF %>% filter(log2FoldChange >= 1) %>% pull(symbol) 

d05_spf <- results.list.comb.filter$d01_d05 %>% filter(log2FoldChange >= 1) %>% pull(symbol) 
```















# GO Terms with clusterProfiler

```{r}
# which ontology: choose between all, biological process (bp), molecular function (mf) 
# and cellular component (cc).
ont <- "bp"

#p-adjust cutoff. Adjustment method is benjamini hochberg (false discovery rate)
pcutoff <- 0.05

go_plots_folder <- file.path(analysis_folder,"GO/plots")
go_charts_folder <- file.path(analysis_folder,"GO/charts")

if (!dir.exists(go_plots_folder)){
  dir.create(go_plots_folder,recursive = T)
  dir.create(go_charts_folder,recursive = T)
}
```








## GO analysis for upregulated Genes (positive values)

```{r}
for(list.nmbr in 1:length(results.list)){
  
GO_NAME <- paste(names(results.list)[list.nmbr],ont,pcutoff,sep = "_")
daypos <- strsplit2(names(results.list[list.nmbr]),split = "_")[2]

print(paste("calculating",daypos,"for",GO_NAME))

res <- results.list[[list.nmbr]]

#UP Regulation Table
sigsdf_UP <- res[res$log2FoldChange >= 1 & (res$padj <= 0.05),]

print(paste("number of genes",nrow(sigsdf_UP)))

UPgene <- sigsdf_UP$symbol
GO_upReg_results <- enrichGO(gene          = UPgene,
                             OrgDb         = "org.Mm.eg.db",
                             keyType       = "SYMBOL",
                             ont           = ont,
                             pAdjustMethod = "BH",
                             pvalueCutoff  = pcutoff,
                             readable      = TRUE)

  #arrange for p-value
  GO_upReg_results@result <- arrange(GO_upReg_results@result,p.adjust)

  GO_upReg_df <- as.data.frame(GO_upReg_results)

  if(nrow(GO_upReg_df) == 0){
    print(paste("No significant GO terms found for",GO_NAME))
    next
    }

  write.csv(GO_upReg_df,file = file.path(go_charts_folder,
                                     glue("GO_{daypos}_results_{GO_NAME}_{condition}.csv")),
            row.names = F)
  
  if(nrow(GO_upReg_df) < 30){
    label_length = nrow(GO_upReg_df)
    } else{label_length = 30}
  
  labels <- str_wrap(rev(GO_upReg_df$Description[1:label_length]),width = 40)
  #dotplot colored by p-value and sorted reverse
  GO_upReg_Dotplot_pvalue <- dotplot.enrichResult2(GO_upReg_results,showCategory = label_length, 
                                                   x = "p.adjust", title = GO_NAME,
                                                   color = "GeneRatio",font.size = 14) +
    scale_x_reverse() + #labels = scaleFUN
    scale_y_discrete(limits = rev(GO_upReg_df$Description[1:label_length]),labels = labels) + 
    ggtitle(glue("GO_{daypos}_dotplot_{GO_NAME}")) + 
    theme(legend.text = element_text(size = 16),legend.title = element_text(size=16),
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
  
  save_ggplot_formats(
    plt = GO_upReg_Dotplot_pvalue,
    create_plot_subdir = T,
    base_plot_dir = go_plots_folder,
    plt_name = glue("GO_{daypos}_Dotplot_pvalue_{GO_NAME}_{condition}"),
    width = 8, height = 12
  )

  #netplot
  GO_upReg_Cnetplot <- cnetplot(GO_upReg_results, 
                                 showCategory = 10,max.overlaps = Inf) +
    ggtitle(glue("GO_{daypos}_Cnetplot_{GO_NAME}_{condition}")) +
    theme(plot.title = element_text(hjust =  0.5,face = "bold",size = 14))

  save_ggplot_formats(
    plt = GO_upReg_Cnetplot,
    create_plot_subdir = T,
    base_plot_dir = go_plots_folder,
    plt_name=glue("GO_{daypos}_Cnetplot_{GO_NAME}_{condition}"),
    width = 12, height = 10
  )
  
  if(nrow(GO_upReg_df) < 5){
    print(paste("Not enough significant GO terms found for treeplot of",GO_NAME))
    next
  }
  #treeplot
  edox2 <- pairwise_termsim(GO_upReg_results)
  GO_upReg_enrichtreeplot <- treeplot(edox2, showCategory = label_length) + 
    ggtitle(glue("GO_{daypos}_enrichtreeplot_{GO_NAME}_{condition}")) + 
    theme(plot.title = element_text(hjust =  0.5,face = "bold",size = 14))
  
  save_ggplot_formats(
    plt = GO_upReg_enrichtreeplot,
    create_plot_subdir = T,
    base_plot_dir = go_plots_folder,
    plt_name = glue("GO_{daypos}_enrichtreeplot_{GO_NAME}_{condition}"),
    width = 15, height = 10
  )
}
```

## GO analysis for DOWNregulated Genes (negative values)

```{r}
for(list.nmbr in 1:length(results.list)){
  GO_NAME <- paste(names(results.list)[list.nmbr],ont,pcutoff,sep = "_")

  dayneg <- strsplit2(names(results.list[list.nmbr]),split = "_")[1]
  
  print(paste("calculating",dayneg,"for",GO_NAME))
  
  res <- results.list[[list.nmbr]]
  
  #DOWN Regulation Table
  sigsdf_DOWN <- res[(res$log2FoldChange <= -1) & (res$padj <= pcutoff),] 
  
  print(paste("number of genes",nrow(sigsdf_DOWN)))
  DOWNgene <- (sigsdf_DOWN$symbol)
  
  GO_downReg_results <- enrichGO(gene = DOWNgene,
                                OrgDb = "org.Mm.eg.db",
                                keyType = "SYMBOL",
                                ont = ont,
                                pAdjustMethod = "BH",
                                pvalueCutoff  = pcutoff,
                                readable      = TRUE)

  #arrange for p-value
  GO_downReg_results@result<-arrange(GO_downReg_results@result,p.adjust)

  GO_downReg_df <- as.data.frame(GO_downReg_results)  
  
  if(nrow(GO_downReg_df) == 0){
    print(paste("No significant GO terms found for",GO_NAME))
    next
  }

  write.csv(GO_downReg_df,file = file.path(go_charts_folder,
              glue("GO_{dayneg}_results_{GO_NAME}_{condition}.csv"))
            ,row.names = F)
  
  if(nrow(GO_downReg_df) < 30){
    label_length = nrow(GO_downReg_df)
    } else{label_length = 30}
  
  labels <- str_wrap(rev(GO_downReg_df$Description[1:label_length]),width = 40)
  
  #dotplot colored by p-value and sorted reverse
  GO_downReg_Dotplot_pvalue <- dotplot.enrichResult2(GO_downReg_results,
                                                     showCategory = label_length, 
                                                    x = "p.adjust", title = GO_NAME,
                          color = "GeneRatio",font.size = 14) + 
    scale_x_reverse() + #labels=scaleFUN
    scale_y_discrete(labels = labels, limits = rev(GO_downReg_df$Description[1:label_length])) +
    ggtitle(glue("GO_{dayneg}_dotplot_{GO_NAME}")) + 
    theme(legend.text = element_text(size = 16),legend.title = element_text(size=16),
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
  
  
  save_ggplot_formats(
    plt = GO_downReg_Dotplot_pvalue,
    create_plot_subdir = T,
    base_plot_dir = go_plots_folder,
    plt_name=glue("GO_{dayneg}_Dotplot_pvalue_{GO_NAME}_{condition}"),
    width = 8, height = 12
  )
 
  #netplot
  GO_downReg_Cnetplot <- cnetplot(GO_downReg_results,
                                 showCategory = 10,max.overlaps = Inf) +
    ggtitle(glue("GO_{dayneg}_Cnetplot_{GO_NAME}")) +
    theme(plot.title = element_text(hjust =  0.5,face = "bold",size = 14))

  save_ggplot_formats(
    plt = GO_downReg_Cnetplot,
    create_plot_subdir = T,
    base_plot_dir = go_plots_folder,
    plt_name=glue("GO_{dayneg}_Cnetplot_{GO_NAME}_{condition}"),
    width = 12, height = 10
  )

  if(nrow(GO_downReg_df) < 5){
    print(paste("Not enough significant GO terms found for treeplot of",GO_NAME))
    next
  }
  #treeplot
  edox2 <- pairwise_termsim(GO_downReg_results)
  GO_downReg_enrichtreeplot <- treeplot(edox2) + 
    ggtitle(glue("GO_{dayneg}_enrichtreeplot_{GO_NAME}_{condition}")) +
    theme(plot.title = element_text(hjust =  0.5,face = "bold",size = 14))
  
  save_ggplot_formats(
    plt = GO_downReg_enrichtreeplot,
    create_plot_subdir = T,
    base_plot_dir = go_plots_folder,
    plt_name=glue("GO_{dayneg}_enrichtreeplot_{GO_NAME}_{condition}"),
    width = 15, height = 10
  )
}
```